src/wallet3/transaction_constructor.cpp

```diff
@@ -29,6 +31,179 @@ namespace wallet
     return new_tx;
   }
 
+  PendingTransaction
+  TransactionConstructor::create_ons_buy_transaction(
+      const cryptonote::tx_destination_entry& change_recipient,
+      const std::string& type_str,
+      const std::string& owner_str,
+      const std::string& backup_owner_str,
+      const std::string& name,
+      const std::string& value
```

Comment:
Some of these arguments are optional, but we're conflating "string is empty" with "value not set", which is not nice from an API point of view.  I think rather the ones that are allowed to be unset should be `std::optional<std::string_view>` arguments.

Reply:
I've updated the function to take std::optional params, however how does an empty param flow through our rpc layer?
```
void RequestHandler::invoke(ONS_BUY_MAPPING& command, rpc_context context) {

  if (auto w = wallet.lock())
  {
    cryptonote::tx_destination_entry change_dest;
    auto ptx = w->tx_constructor->create_ons_buy_transaction(
      command.request.name,
      command.request.type,
      command.request.value,
      command.request.owner,
      command.request.backup_owner,
      change_dest
      );

...
    command.response["status"] = "200";
    command.response["result"] = submit_future.get();
  }
}
```
Is command.request going to pass us an empty string anyway?

Reply:
Those should be `std::optional<std::string>`s too.